internal/certgen/certgen.go

```diff
@@ -39,107 +55,107 @@ func writePEM(outputDir, filename string, data []byte) error {
-	if err != nil {
-		return err
-	}
-	return writePEM(outputDir, "envoykey.pem", certdata["envoykey.pem"])
 
-}
-
-// WriteSecretsYAML writes all the keypairs out to Kube Secrets in YAML form
-// in outputDir. The CA Secret only contains the cert.
-func WriteSecretsYAML(outputDir, namespace string, certdata map[string][]byte) error {
-	err := writeCACertSecret(outputDir, "cacert.pem", certdata["cacert.pem"])
-	if err != nil {
-		return err
-	}
-	err = writeKeyPairSecret(outputDir, "contour", namespace, certdata["contourcert.pem"], certdata["contourkey.pem"])
+	err = writePEM(outputDir, "contourcert.pem", certdata[ContourCertificateKey], force)
 	if err != nil {
 		return err
 	}
 
-	return writeKeyPairSecret(outputDir, "envoy", namespace, certdata["envoycert.pem"], certdata["envoykey.pem"])
-
-}
-
-// WriteSecretsKube writes all the keypairs out to Kube Secrets in the
-// passed Kube context.
-func WriteSecretsKube(client *kubernetes.Clientset, namespace string, certdata map[string][]byte) error {
-	err := writeCACertKube(client, namespace, certdata["cacert.pem"])
+	err = writePEM(outputDir, "contourkey.pem", certdata[ContourPrivateKeyKey], force)
 	if err != nil {
 		return err
 	}
-	err = writeKeyPairKube(client, "contour", namespace, certdata["contourcert.pem"], certdata["contourkey.pem"])
+
+	err = writePEM(outputDir, "envoycert.pem", certdata[EnvoyCertificateKey], force)
 	if err != nil {
 		return err
 	}
 
-	return writeKeyPairKube(client, "envoy", namespace, certdata["envoycert.pem"], certdata["envoykey.pem"])
-
+	return writePEM(outputDir, "envoykey.pem", certdata[EnvoyPrivateKeyKey], force)
 }
 
-func writeCACertSecret(outputDir, namespace string, cert []byte) error {
-	filename := path.Join(outputDir, "cacert.yaml")
-	secret := newCertOnlySecret("cacert", namespace, "cacert.pem", cert)
-	f, err := createFile(filename, false)
-	if err != nil {
-		return err
+// WriteSecretsYAML writes all the keypairs out to Kubernetes Secrets in YAML form
+// in outputDir.
+func WriteSecretsYAML(outputDir string, secrets []*corev1.Secret, force OverwritePolicy) error {
+	for _, s := range secrets {
+		filename := path.Join(outputDir, s.Name+".yaml")
+		f, err := createFile(filename, force == Overwrite)
+		if err != nil {
+			return err
+		}
+		if err := checkFile(filename, writeSecret(f, s)); err != nil {
+			return err
+		}
 	}
-	return checkFile(filename, writeSecret(f, secret))
+
+	return nil
 }
 
-func writeCACertKube(client *kubernetes.Clientset, namespace string, cert []byte) error {
-	secret := newCertOnlySecret("cacert", namespace, "cacert.pem", cert)
-	if _, err := client.CoreV1().Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
-		if k8serrors.IsAlreadyExists(err) {
-			fmt.Print("secret/cacert already exists\n")
-			return nil
+// WriteSecretsKube writes all the keypairs out to Kubernetes Secrets in the
+// compact format which is compatible with Secrets generated by cert-manager.
```

Comment:
If cert-manager is being used, why have this in the mix? Shouldn't cert-manager handle all the cert generation?

Reply:
It's so that you can start with `certgen` and then change your mind and move to cert-manager, or vice versa.

Reply:
Yup; for general interoperability without having to fork and own the deployment YAML.

Reply:
Sorry, I now become bit confused about this topic in general :)  So I'd would like to ask you if this reasoning holds:

* The default example deployment should start to use `contour certgen` "compat" option - The motivation is to allow converting any new Contour deployment to cert-manager easily.
* The "legacy" option is still default behavior for `contour certgen` - The motivation is to allow people to upgrade their existing deployments by just changing the version tag.

On the other hand, all existing deployments already have their legacy secrets generated previously.  So now my confusion is: what is the scenario when new version of `contour certgen` needs to generate **old** secret?

Isn't the backwards compatibility problem rather coming from the changes needed in volume mounts and command line parameters (i.e. the last four bullets in https://github.com/projectcontour/contour/issues/2149#issuecomment-598065053)?

I guess this question could potentially be quite fundamental for this PR, sorry for being late on this!!!

Reply:
The main motivation for adding the new flags is so that we can make the configuration change separately (and revert it if I screw it up). After this PR, the default `certgen` behavior will be unchanged, so I can then make a separate PR to enable the new options along with the deployment changes. This way, the repo moves from one consistent state to another, and we can safely stop at any time.